lib/base/io-engine.hpp

```diff
@@ -172,14 +174,16 @@ class Timeout : public SharedObject
 
 	template<class Executor, class TimeoutFromNow, class OnTimeout>
 	Timeout(boost::asio::io_context& io, Executor& executor, TimeoutFromNow timeoutFromNow, OnTimeout onTimeout)
-		: m_Timer(io)
+		: m_Timer(io), m_Done(io)
 	{
 		Ptr keepAlive (this);
 
 		m_Cancelled.store(false);
 		m_Timer.expires_from_now(std::move(timeoutFromNow));
 
-		IoEngine::SpawnCoroutine(executor, [this, keepAlive, onTimeout](boost::asio::yield_context yc) {
+		auto setDone (Shared<Defer>::Make([this, keepAlive] { m_Done.Set(); }));
```

Comment:
What's the reason for this shared pointer instead of doing this inside the lambda just below?